### PR TITLE
Fixes Spontaneous Harddel with Boned Fish Revive Ability

### PR DESCRIPTION
**Pull request by san7890**
## About The Pull Request

You can find an example of the failing stack trace here: https://github.com/tgstation/tgstation/actions/runs/8242191499/job/22540786029?pr=81937#step:10:1323

Here's a screenshot:

image

Explanation: The boned fish's revive ability added a timer that would randomly fire in the next one-to-two minutes. Due to this inherent randomness, create-and-destroy would locate the fact that we were hanging refs to this fish since the timer may not have processed by the time the fish was deleted (probably if the timer was considerably more than one minute), so in order to fix this let's just add the `TIMER_DELETE_ME` flag to the revive ability so that the timer get's `qdel`'d on the fish's `Destroy()`. simple

i probably cocked up some detail of the above explanation but the timer flag was literally created to prevent stuff like this from happening so let's use it

## Changelog

don't matter

### CODE REVIEW

**Comment by MrMelbert** [comment]
> let's just add the TIMER_DELETE_ME flag to the revive ability so that the timer get's qdel'd on the fish's Destroy(). simple

Are you sure this'll work? The timer isn't on the fish itself, it's on the `/datum/fish_trait` singleton(?), so `TIMER_DELETE_ME` shouldn't be doing anything

**Comment by san7890** [comment]
> Are you sure this'll work? The timer isn't on the fish itself, it's on the `/datum/fish_trait` singleton(?), so `TIMER_DELETE_ME` shouldn't be doing anything

that's a good point. i'll have the fish trait datum qdel itself if the fish gets destroyed

**Comment by MrMelbert** [comment]
> that's a good point. i'll have the fish trait datum qdel itself if the fish gets destroyed

I think they're singletons. Emphasis on "I think".

**Comment by san7890** [comment]
alright it was a bit more wieldy to handle it with the singleton consideration (you were right) but i think the solution should work:tm:

**Comment by MrMelbert** [comment]
I'd almost think it would've been easier to just put the `revive` proc onto `/obj/item/fish` itself and resort to the original method (of using `DELETE_ME`

**Comment by san7890** [comment]
i was pondering involving `/obj/item/fish` but all of the code I saw re: fish traits is that it tended to be pretty modularized and not really involve the fish itself (it doesn't even hold a ref to the fish) so I decided against it. i don't think it would be any cleaner to fracture the code like that

File: code/modules/fishing/fish/fish_traits.dm [hunk 1]
```diff
@@ -206,16 +206,23 @@ GLOBAL_LIST_INIT(fish_traits, init_subtypes_w_path_keys(/datum/fish_trait, list(
...
 	RegisterSignal(fish, COMSIG_FISH_STATUS_CHANGED, PROC_REF(check_status))
 
 /datum/fish_trait/revival/proc/check_status(obj/item/fish/source)
 	SIGNAL_HANDLER
-	if(source.status == FISH_DEAD)
-		addtimer(CALLBACK(src, PROC_REF(revive), source), rand(1 MINUTES, 2 MINUTES))
+	if(source.status != FISH_DEAD)
+		return
+
+	var/timer_id = addtimer(CALLBACK(src, PROC_REF(revive), source), rand(1 MINUTES, 2 MINUTES), flags = TIMER_STOPPABLE)
```
Comment (LemonInTheDark) [inline code/modules/fishing/fish/fish_traits.dm:220]: why not just use a weakref?

**Comment by LemonInTheDark** [comment]
This is an annoying problem, singletons are truly the devil

File: code/modules/fishing/fish/fish_traits.dm (hunk 1 above)
Comment (san7890) [reply code/modules/fishing/fish/fish_traits.dm:220]: i was pondering using that but i figured it wasn't necessary for what we needed to accomplish here (which is just store a list of strings (timer ids), and things to look up those strings)

Comment (MrMelbert) [reply code/modules/fishing/fish/fish_traits.dm:220]: I think they mean "why not pass the fish-to-be-revived via weakref in the callback", I think that would work too

Comment (san7890) [reply code/modules/fishing/fish/fish_traits.dm:220]: oh i see, i think this is probably the most elegant solution

**Comment by san7890** [comment]
ignore that i cocked up the order of operations

**Comment by san7890** [comment]
I think this PR is in a much better spot with the current implementation than when Melbert made that suggestion (see latest code comment review)

**Review by Ghommie — APPROVED** [review]

**Pull request merged**